RP2040 MCU Support

vxj9800 opened the pull request:
Submitting a pull as asked by the dev-team during [PX4 dev call](https://discuss.px4.io/t/px4-dev-call-july-21-2021/23297).

**Describe problem solved by this pull request**
Add PX4 base layer support for RP2040 MCU

**Describe your solution**
Currently, this branch is created from @davids5's  [pr-nuttx-10.1.0+-upstream](https://github.com/PX4/PX4-Autopilot/tree/pr-nuttx-10.1.0+-upstream) branch because the nuttx in current master doesn't have support for RP2040. Let me know if I should do thing differently in this matter.
So far, files with absolute necessity are added so that PX4 compiles. Other base layer functions for adc, spi, etc. will be added eventually.

**Test data / coverage**
Working on it.

**Additional context**
Creating this pull request as a draft because many more changes will happen before the PX4 base layer is ready for RP2040. The main goal of creating this PR is to get some feedback from dev-team.

Please take a look at commits made since July 14th.

dagar commented:
I've changed the base branch of the PR to `pr-nuttx-10.1.0+-upstream` to try and minimize the diff, but it looks like you're slightly behind? Could you rebase on current `pr-nuttx-10.1.0+-upstream`?

vxj9800 commented:
> I've changed the base branch of the PR to `pr-nuttx-10.1.0+-upstream` to try and minimize the diff, but it looks like you're slightly behind? Could you rebase on current `pr-nuttx-10.1.0+-upstream`?

@dagar , Sorry to ask you this but do you want me to something like [this](https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a). I am still learning git. Sorry.

dagar commented:
> > I've changed the base branch of the PR to `pr-nuttx-10.1.0+-upstream` to try and minimize the diff, but it looks like you're slightly behind? Could you rebase on current `pr-nuttx-10.1.0+-upstream`?
> 
> @dagar , Sorry to ask you this but do you want me to something like [this](https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a). I am still learning git. Sorry.

Yes exactly. In this case you can do `git rebase upstream/pr-nuttx-10.1.0+-upstream`.

vxj9800 commented:
> > > I've changed the base branch of the PR to `pr-nuttx-10.1.0+-upstream` to try and minimize the diff, but it looks like you're slightly behind? Could you rebase on current `pr-nuttx-10.1.0+-upstream`?
> > 
> > 
> > @dagar , Sorry to ask you this but do you want me to something like [this](https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a). I am still learning git. Sorry.
> 
> Yes exactly. In this case you can do `git rebase upstream/pr-nuttx-10.1.0+-upstream`.

I have rebased to current pr-nuttx-10.1.0+-upstream. Please let me know if I did anything wrong.

davids5 commented on boards/myboard/myfc/default.cmake:
```diff
@@ -0,0 +1,105 @@
+px4_add_board(
+PLATFORM nuttx
+VENDOR myboard
+MODEL myfc
```
We will need to nail these done at some point (The ripple can be large so sooner is better)

dir names comments and firmware.prototype and booloader QGC etc will have to change

davids5 commented on boards/myboard/myfc/default.cmake:
```diff
@@ -0,0 +1,105 @@
+# 	usb_connected
+# 	ver
+# 	work_queue
+)
```
There should be some simple things that can be enabled above.

[this board](https://www.seeedstudio.com/ARDUINO-NANO-RP2040-CONNECT-WITH-HEADERS-p-4941.html?gclid=Cj0KCQjw0emHBhC1ARIsAL1QGNf-GtXTRIa8gkXtLtUyuvqL9llB5J3dkLwOqNLLiq6Hgsa44-5NNVYaAqotEALw_wcB) may be a good port to nuttx to get some more feature support. It has an IMU.... 


Some above are hardware dependent, but look for easy wins and it will help create a todo list for nuttx driover support. 

E.G. SD support over SPI. use sd_bench to test.

UUID may be solved with something like https://www.microchip.com/wwwproducts/en/24aa02uid (bigger will be better)  This can also be used to store a small amount of parameters for the param system (Flash can be used as well) but that will be need a program mem nuttx driver. 

ver, top serial_test  should be easy wins.

davids5 commented on boards/myboard/myfc/nuttx-config/nsh/defconfig,old:
```diff
@@ -0,0 +1,54 @@
+CONFIG_SYSTEM_NSH=y
+CONFIG_USBDEV=y
+CONFIG_USBDEV_BUSPOWERED=y
+CONFIG_USER_ENTRYPOINT="nsh_main"
```
Delete this file, git will do the history

davids5 commented on boards/myboard/myfc/src/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+add_library(drivers_board
+	# i2c.cpp
```
I see 12c, serial and spin in NuttX https://github.com/apache/incubator-nuttx/tree/master/arch/arm/src/rp2040

Assuming they are not stubbed out there, these can came in 

I2c in nuttx already has reference counting to that is good to go! 
Tad bit of caution the CONFIG_I2C_RESET is not really supported (it is currently a nop in nuttx)

davids5 commented on boards/myboard/myfc/src/init.c:
```diff
@@ -0,0 +1,409 @@
+ ************************************************************************************/
+__EXPORT void board_on_reset(int status)
+{
+	// /* configure the GPIO pins to outputs and keep them low */
```
Looking at the GPIO: You will need to create shims in the micro hall to support the PX4 px4_arch_XXXXgpio API. 

This is not needed in arch dependent board files but any other code that does GPIO in the PX4 layer will need it. 

See how the [Kinetis HW  platforms/nuttx/src/px4](https://github.com/PX4/PX4-Autopilot/blob/master/platforms/nuttx/src/px4/nxp/kinetis/io_pins/kinetis_pinirq.c) creates kinetis_gpiosetevent then it is mapped to [px4_arch_gpiosetevent](https://github.com/PX4/PX4-Autopilot/blob/master/platforms/nuttx/src/px4/nxp/k66/include/px4_arch/micro_hal.h#L112)

davids5 reviewed (commented):
This is coming along Good work!

The rebase looks OK. Some note are below.

The PX4 platform needs a HRT, Output (PWM), Serial (console, GPS)m and some sensors (SPI preferred for IMU) Slow device BARO, Compass: ONBOARD I2C is OK. (Long I2C runs CAN be a real problem: and CAN is the solution (pun intended)

Parameter need to in NON-Vol storage.
Logging can be streamed but SD card (SPI can be used albeit it sub optimize speed wise) is more manageable if data links do not have good QOS or are fractional. 

LED's Tones Button are nice to have (tones help validate the timers :)

Hard fault logging can be done without BBSRAM is is just more prone to being not as reliable as having HW write protection.
UUID (see suggestion) Needed at some point, but can hard coded for DEV

vxj9800 commented:
> This is coming along Good work!
> 
> The rebase looks OK. Some note are below.
> 
> The PX4 platform needs a HRT, Output (PWM), Serial (console, GPS)m and some sensors (SPI preferred for IMU) Slow device BARO, Compass: ONBOARD I2C is OK. (Long I2C runs CAN be a real problem: and CAN is the solution (pun intended)
> 
> Parameter need to in NON-Vol storage.
> Logging can be streamed but SD card (SPI can be used albeit it sub optimize speed wise) is more manageable if data links do not have good QOS or are fractional.
> 
> LED's Tones Button are nice to have (tones help validate the timers :)
> 
> Hard fault logging can be done without BBSRAM is is just more prone to being not as reliable as having HW write protection.
> UUID (see suggestion) Needed at some point, but can hard coded for DEV

Thanks for the review @davids5. I have implemented I2c, SPI, serial and some other things that you recommended. I wanted to start testing the build on the hardware. Unfortunately, I have hit a dead end in this matter.

The problem is that the RP2040 has a two stage boot sequence. Nuttx uses some files form [pico-sdk](https://github.com/raspberrypi/pico-sdk), some other files located at [nuttx/tools/rp2040](https://github.com/PX4/NuttX/tree/30135b5f1879c0e5625e410a8437395d3bd78e30/tools/rp2040) and at [nuttx/arch/arm/src/rp2040/boot2](https://github.com/PX4/NuttX/tree/30135b5f1879c0e5625e410a8437395d3bd78e30/arch/arm/src/rp2040/boot2) to accommodate this during compilation. The general compilation process for a specific board in nuttx is handled by [Make.defs](https://github.com/PX4/NuttX/blob/30135b5f1879c0e5625e410a8437395d3bd78e30/boards/arm/rp2040/raspberrypi-pico/scripts/Make.defs) file in the scripts directory of the board. The PX4 doesn't use this procedure, as far as I understand. Is there a quick solution to this problem in PX4?

dagar commented:
> The problem is that the RP2040 has a two stage boot sequence. Nuttx uses some files form pico-sdk, some other files located at nuttx/tools/rp2040 and at nuttx/arch/arm/src/rp2040/boot2 to accommodate this during compilation. The general compilation process for a specific board in nuttx is handled by Make.defs file in the scripts directory of the board. The PX4 doesn't use this procedure, as far as I understand. Is there a quick solution to this problem in PX4?

PX4 wraps the regular NuttX Makefile build. https://github.com/PX4/PX4-Autopilot/blob/master/platforms/nuttx/NuttX/Make.defs.in


Can you get standalone NuttX working entirely on this hardware? I can take a look at how to adapt that part of the NuttX build to cleanly fit within what PX4 calls.

vxj9800 commented:
> > The problem is that the RP2040 has a two stage boot sequence. Nuttx uses some files form pico-sdk, some other files located at nuttx/tools/rp2040 and at nuttx/arch/arm/src/rp2040/boot2 to accommodate this during compilation. The general compilation process for a specific board in nuttx is handled by Make.defs file in the scripts directory of the board. The PX4 doesn't use this procedure, as far as I understand. Is there a quick solution to this problem in PX4?
> 
> PX4 wraps the regular NuttX Makefile build. https://github.com/PX4/PX4-Autopilot/blob/master/platforms/nuttx/NuttX/Make.defs.in
> 
> Can you get standalone NuttX working entirely on this hardware? I can take a look at how to adapt that part of the NuttX build to cleanly fit within what PX4 calls.

@dagar , Yes, the standalone Nuttx  works perfectly fine on raspberrypi-pico (which I am also using as a development board currently). I have tested this by running [usbnsh](NuttX/boards/arm/rp2040/raspberrypi-pico/configs/usbnsh/) configuration on the pico.

I guess following two files adds extra stuff to the regular Nuttx compilation process for RP2040.
[NuttX/arch/arm/src/rp2040/boot2/Make.defs](https://github.com/PX4/NuttX/blob/8660572a3c69e0e2ca24c3cabeaaaae3372a6960/arch/arm/src/rp2040/boot2/Make.defs)
[NuttX/tools/rp2040/Config.mk](https://github.com/PX4/NuttX/blob/8660572a3c69e0e2ca24c3cabeaaaae3372a6960/tools/rp2040/Config.mk)

The Make.defs above seems to be the more important one as it uses some files from pico-sdk. The Config.mk from tools folder seems to just convert .elf file to .uf2 file which can directly be copied in the pico during usb mass storage mode. This is not very important as this conversion can be done very easily using elf2uf2 program provided in pico-sdk.

davids5 commented:
I have merged PX4:pr-nuttx-10.1.0+-ustream into PX4:pr-nuttx-10.1.0+ and will merge to master today or tomorrow. This can be rebased on that and then reopened.

vxj9800 commented:
> I have merged PX4:pr-nuttx-10.1.0+-ustream into PX4:pr-nuttx-10.1.0+ and will merge to master today or tomorrow. This can be rebased on that and then reopened.

Hello @davids5, I have rebased my branch to the current master. Could you please reopen the PR? Thanks.

dagar commented:
@vxj9800 it appears that github won't let us reopen this PR, can you open a fresh one?

vxj9800 commented:
@dagar In that case, will it be beneficial to create a new branch from PX4\PX4-Autopilot master and then move everything from my fork to the new branch before opening a new PR?

davids5 commented:
@vxj9800 

If you rebase locally on master. Then `git branch -m rp2040-support pr-rp2040-support` you should be able to push that and open a new PR against PX4 master

vxj9800 commented:
@davids5 
I tried your suggestion. I got an error saying that I don't have permission. So, I guess I either need permission or I messed up in the process somewhere.
